### PR TITLE
Add Stop Primitive Classes

### PR DESCRIPTION
**Pull request by jared-paul**
### Description

Adds the stop primitive classes

### Testing Done

Compiled successfully with the code, initialized constructor successfully

### Resolved Issues

#90 

### Review Checklist

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

File: src/thunderbots/software/ai/primitive/stop_primitive.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.cpp]: Github complaining about lack of newline at end of file.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: Github complaining about lack of newline at end of file.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h:12]: It's helpful to also include a quick description of what the primitive does, eg "stops the robot in place, with an option to coast rather than stop"

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: I _think_ this boolean parameter is to enable coasting rather than enable stopping. I'm not somewhere where I can check rn, but that's also what we have in our primitive specs from #86. We should double check that and change it if necessary.

@henrymliu fyi

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just a few small things otherwise looks good!

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (henryymliu) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: Yes that's correct, one value coasts, the other applies reverse voltage to the motors
Can't remember which one is which

**Comment by henryymliu** [comment]
After addressing Mathew's comments, this looks good to me.

**Comment by garethellis0** [comment]
Only thing here is that we're missing unit tests. @MathewMacDougall - did we want to wait until https://github.com/UBC-Thunderbots/Software/issues/129 is done before adding tests to all the primitives? Or did we want to extend that issue to also cover adding unit test to _all_primitives and just merging the ones we have in open PR's without testing?

**Comment by MathewMacDougall** [comment]
Since #129 is blocked by #9, lets extend #129 to cover testing _all_ `Primitive` classes and merge the existing PR's without testing.

**Comment by MathewMacDougall** [comment]
@jared-paul it would be great if you could also add some unit tests for this class as well. That will give us test coverage right away, and is a good introduction to testing in our codebase.

You can use #131 as an example. Sorry for the last minute change.

**Comment by MathewMacDougall** [comment]
@jared-paul ETA on unit tests?

**Comment by MathewMacDougall** [comment]
@jared-paul any updates? It would be great if we could have this done by Christmas

File: src/thunderbots/software/ai/primitive/stop_primitive.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.cpp]: Github complaining about lack of newline at end of file.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h:12]: It's helpful to also include a quick description of what the primitive does, eg "stops the robot in place, with an option to coast to a stop rather than stop immediately"

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: Github complaining about lack of newline at end of file.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: This function should have a javadoc comment explaining what it does

File: src/thunderbots/software/test/primitive/stop_primitive.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/test/primitive/stop_primitive.cpp]: `
```suggestion
 * This file contains the unit tests for the StopPrimitive class
```

File: src/thunderbots/software/test/primitive/stop_primitive.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/test/primitive/stop_primitive.cpp]: Replace `int()` with `0`. It's a bit more explicit and matches the constructor above

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Overall looks good to me, just a few nits about comments and newlines.

@jared-paul could you add one more test in the `test/primitive.cpp` file, for converting a Primitive to a message, and then back to a Primitive? You can create a new test called `convert_StopPrimitive_to_message_and_back_to_StopPrimitive`, and base it off of `convert_MovePrimitive_to_message_and_back_to_MovePrimitive`

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: Could we rename this to `robotShouldCoast()`? I just think it's a bit more descriptive (it's describing what the robot should do, rather then some state of the robot, which I think `coastRobot` implies).

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: `getParameters` and `getExtraBits` should have descriptive javadoc comments, see other primitives for examples

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: class members should each have a quick descriptive comment above them.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking pretty good!

- we should add a test for converting a primitive to a message and back in `test/primitive/primitive.cpp` (see tests for other primitives in that file for examples)

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: I don't think it needs one. This is overriding a function, and the commend documentation is there. We typically haven't been re-commenting overridden functions.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: I would say yes `coast` should have one, but I think comments for `robot_id` is redundant since it's already a descriptive variable name.

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: Generally I agree, but I think it's important that these functions have explicit documentation for the form of the vector that they're returning (or if those vectors are empty) because it's otherwise not indicated what the semantic meaning of each element in the vector means. See other primitives for examples.

Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: Ok fair point.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:. Nice work!

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Whoops missed the CI check. Looks like you'll need to fix the code formatting. Run `./clang_format/fix_formatting.sh -a` to automatically fix the code formatting. Then you can just make another commit called something like "fixed formatting"

You'll also need to update your branch to get the latest changes. You can run `git pull upstream master`.

**Review by MathewMacDougall — DISMISSED** [review]

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (garethellis0) [inline src/thunderbots/software/ai/primitive/stop_primitive.h]: nit; but we can change this to just 
```
     // whether the robot should coast or not
````

Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: Also, what does it mean to "coast" here? It would be great if this comment somewhat explained that (just at a high level).

**Comment by garethellis0** [comment]
as a quick note for the future, could you please leave a quick comment reply to each of my comments and leave them unresolved? The comment reply can be as simple as "resolved", or "not resolved, because ..." and leaving them unresolved makes it easy for me to come back and give myself context for this PR (especially when I'm reviewing multiple PR's for multiple rounds). Thanks!

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: If we're going to make it `//` might as well put it on the line directly above the variable.

`coast` means the robot will not apply power to its wheels and slow down to a stop naturally. Stopping without coasting will make the robot actively try stop in place, and therefore stop sooner.

**Comment by MathewMacDougall** [comment]
@jared-paul how's progress on the latest comments?

**Comment by jared-paul** [comment]
My project kind of broke when I tried to update my branch with "git pull upstream master" so it deleted some of my code when I had to overwrite my version with the repos, my latest commit reflects that.

And sorry about the whole review process, I don't really know how github pull requests/reviews work

**Comment by garethellis0** [comment]
~@jared-paul - could you pull the latest commits from `master` of this repo to your branch (ie. update `master` on your fork, then merge `master` into this branch)? I think that will fix the kinda funky codecov.~

~Let me know if you need any help with this!~

**EDIT:** Sorry, I misunderstood what the issue was............. could you please rebase your branch to master (on your branch, run `git rebase master`)? This will ensure that the pull request base (the last commit that this branch and master have in common) has the code coverage stuff setup on it. Again, lemme know if you need any help with this.

**Comment by MathewMacDougall** [comment]
That's weird that your changes would have been removed. `git pull` shouldn't overwrite your code like that. If that happens again let me know and we can try figure out what's going on.

@garethellis0 his branch isn't being reported as out of date, so not sure if another pull will fix coverage?

**Comment by garethellis0** [comment]
Yeah, another pull won't, but if you rebase it will update the commit at which this branch and master diverge (the base) and that's what codecov uses to diff (from base to head) to determine changes in coverage.

**Comment by jared-paul** [comment]
When I run "git pull upstream master" or "git rebase master" it says "already up to date"
http://prntscr.com/m56bjl

**Comment by MathewMacDougall** [comment]
@garethellis0 I'm not sure rebasing works very nice once there are merge commits in your history since the branch diverged.

@jared-paul try running `git pull --rebase upstream master`. `git rebase master` will try rebase onto your _local_ `master` branch, but we want to rebase on the _upstream_ `master` branch

**Comment by garethellis0** [comment]
Hmm, no worries if it doesn't work once you try Mat's suggestion.

**Comment by jared-paul** [comment]
When I ran "git pull --rebase upstream master" it said http://prntscr.com/m56ojf
It also added a whole bunch of weird things to my code like http://prntscr.com/m56owu

**Comment by MathewMacDougall** [comment]
@jared-paul Yup that's expected. Those weird things in the code appear when a merge conflict occurs that you have to resolve. Lets move conversation to Slack since communication should be quicker there.

**Comment by MathewMacDougall** [comment]
I don't see a new coverage report. @garethellis0 can we not trigger a new report manually?

**Comment by garethellis0** [comment]
Hmm, there is no coverage report for the two most recent commits on github. I'm not sure what (if anything) changed, but I've re-triggered the build for the most recent commit to see if it fixes it. Either way we should be able to merge this, and since coverage has clearly run for this PR, hopefully that should mean that coverage will start showing up on master again. Please let me know if you see any more occurrences of this once this PR is merged!

**Review by MathewMacDougall — DISMISSED** [review]
Sounds good

**Comment by MathewMacDougall** [comment]
@jared-paul you'll need to update your branch but otherwise lets go ahead and merge this. Coverage should be fine.

File: src/thunderbots/software/ai/primitive/stop_primitive.cpp
Comment (henryymliu) [reply src/thunderbots/software/ai/primitive/stop_primitive.cpp]: bump

File: src/thunderbots/software/ai/primitive/stop_primitive.h
Comment (henryymliu) [reply src/thunderbots/software/ai/primitive/stop_primitive.h]: bump?

**Review by henryymliu — CHANGES_REQUESTED** [review]
Literally two new lines, and we're done :+1:
(not sure if they were accidentally removed beforehand)

**Comment by MathewMacDougall** [comment]
@jared-paul how's progress on the last changes? This is super-close to being done!

**Review by henryymliu — APPROVED** [review]
LGTM :+1:

**Review by MathewMacDougall — APPROVED** [review]
LGTM :tada:

**Comment by MathewMacDougall** [comment]
@garethellis0 please check that your changes were addressed

**Review by garethellis0 — APPROVED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
@jared-paul You'll just need to update your branch before you can merge. `git pull upstream master` should do it (make sure to run the formatting script afterwards, just in case. `clang_format/fix_formatting.sh -a`)

**Comment by MathewMacDougall** [comment]
@jared-paul any updates? Would be good to quickly get this done before I put another PR up that will cause this one to need to be modified.

**Comment by MathewMacDougall** [comment]
Looks good to go :+1:

**Pull request merged**
